File: app/api/routes/follows.py
from app.models import db
from app.models.follows import Follow
from flask import Blueprint, jsonify, request
from flask_login import current_user, login_required

follow = Blueprint("follows", __name__)


@follow.route("/<int:userId>")
@login_required
def follows_list(userId):
    """
    Query for the users follows by ID

    if the current user is matches the user id - return data for that user Id
    """

    following = Follow.query.filter(Follow.user_id == int(userId)).all()
    followers = Follow.query.filter(Follow.followed_user_id == int(userId)).all()

    final_obj = {
        "Following": [follow.to_dict() for follow in following],
        "Followers": [follower.to_dict() for follower in followers],
    }

    # print(' FOLLOWS HERE ===>  ', following)

    return final_obj


@follow.route("/<int:userId>/<int:otherUserId>/follow", methods=["POST"])
@login_required
def follow_other_user(userId, otherUserId):
    # * query the db for the user following otherUser

    request_data = request.get_json()
    # print('THIS IS REQUEST DATA ===>  ', request_data)
    foundFollow = Follow.query.filter(
        Follow.user_id == userId, Follow.followed_user_id == otherUserId
    ).first()
    # print('DID WE FIND THE FOLLOW?')
    # print('          !!!!!!!!!  ', foundFollow)

    if not foundFollow:
        new_follow = Follow(
            user_id=userId,
            user_username=request_data["mainUserUsername"],
            followed_user_id=otherUserId,
            followed_user_username=request_data["otherUserUsername"],
        )

        # print('            !!!!!!!!!!! NEW FOLLOW ===>   ', new_follow.user_id, new_follow.followed_user_id)

        db.session.add(new_follow)
        db.session.commit()
        following = Follow.query.filter(Follow.user_id == int(otherUserId)).all()
        followers = Follow.query.filter(
            Follow.followed_user_id == int(otherUserId)
        ).all()

        final_obj = {
            "Following": [follow.to_dict() for follow in following],
            "Followers": [follower.to_dict() for follower in followers],
        }

        return final_obj
    return jsonify({"errors": "User not found"})


@follow.route("/<int:userId>/<int:otherUserId>/unfollow", methods=["DELETE"])
@login_required
def unfollow_other_user(userId, otherUserId):
    """
    Query for the follow row in the db, unfollow the other user,
    and return the updated other user's follows
    """
    request_data = request.get_json()
    print("THIS IS REQUEST DATA ===>  ", request_data)
    found_follow = Follow.query.filter(
        Follow.user_id == userId, Follow.followed_user_id == otherUserId
    ).first()
    # print('DID WE FIND THE FOLLOW?')
    # print('          !!!!!!!!!  ', found_follow)

    if found_follow:
        db.session.delete(found_follow)
        db.session.commit()

        following = Follow.query.filter(Follow.user_id == int(otherUserId)).all()
        followers = Follow.query.filter(
            Follow.followed_user_id == int(otherUserId)
        ).all()

        final_obj = {
            "Following": [follow.to_dict() for follow in following],
            "Followers": [follower.to_dict() for follower in followers],
        }

        return final_obj

    return jsonify({"errors": "User not found"})


# User goes to friends/own profile page
# 1. follows object gets loaded for the user in the url params
# - frontend sends a request to the redux thunk to load the follows object
# - redux thunk fetches backend data from user follows with the userId
# - backend sends follows object with "Followers" & "Following" keys
# 2. everytime the user page is rendered (whether current user or other user)
# the follows fetch is made to backend
# Add editing logic in frontend using the url params to check for userId

File: react-vite/src/redux/follows.js
const LOAD_FOLLOWS = "follows/loadFollows";
const FOLLOW_OTHER_USER = "follows/followOtherUser"
const UNFOLLOW_OTHER_USER = "follows/unfollowOtherUser"

const loadFollows = (payload) => ({
  type: LOAD_FOLLOWS,
  payload,
});
const followOtherUserAction = (payload) => ({
  type: FOLLOW_OTHER_USER,
  payload,
});
const unfollowOtherUserAction = (payload) => ({
  type: UNFOLLOW_OTHER_USER,
  payload,
});

export const thunkLoadFollows = (userId) => async (dispatch) => {
  // console.log("        IN LOAD FOLLOWS THUNK ===>  ", userId);

    
    const response = await fetch(`/api/follows/${userId}`, {
      method: "GET",
      headers: { "Content-Type": "application/json" },
    });

    if (response.ok) {
      const data = await response.json();
      // console.log('DATA HERE FROM FOLLOWS ==>  ', data);

      await dispatch(loadFollows(data));
    }
    return response;
 
};
export const thunkFollowOtherUser = (userId, mainUserUsername, otherUserId,otherUserUsername) => async (dispatch) => {
console.log('IN FOLLOW OTHER USER THUNK ==>');
// ** GOAL: return the follows for the Other User after the main user follows them
// ! ONLY WANT TO RETURN THE FOLLOWERS STATE TO UPDATE FOR THE OTHER USER
const response = await fetch(`/api/follows/${userId}/${otherUserId}/follow`, {
  method: 'POST',
  headers: { "Content-Type": "application/json" },
  body: JSON.stringify({
    mainUserUsername,
    otherUserUsername
  })
})
const data = await response.json();
if(data.errors){
  return data.errors
}

if(response.ok){
  console.log('FOLLOW DATA THUNK ==>    ', data);
  dispatch(followOtherUserAction(data))
}
return response;
};


export const thunkUnfollowOtherUser = (userId, mainUserUsername, otherUserId,otherUserUsername) => async (dispatch) => {
  console.log(' IN UNFOLLOW OTHER USER THUNK ==>');
  // ** GOAL: return the follows for the Other User after the main user unfollows them
  // ! ONLY WANT TO RETURN THE FOLLOWERS STATE TO UPDATE FOR THE OTHER USER
  const response = await fetch(`/api/follows/${userId}/${otherUserId}/unfollow`, {
    method: 'DELETE',
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify({
      mainUserUsername,
      otherUserUsername
    })
  })
  
  const data = await response.json();
  if(data.errors){
    return data.errors
  }

  if(response.ok){
    console.log('UNFOLLOW DATA THUNK ==>    ', data);
    dispatch(unfollowOtherUserAction(data));
  }

  return response
};



const initialState = {};

function followsReducer(state = initialState, action) {
  switch (action.type) {
    case LOAD_FOLLOWS: {
      return { ...state, ...action.payload };
    }
    case FOLLOW_OTHER_USER: {
      return { ...state, ...action.payload };
    }
    case UNFOLLOW_OTHER_USER: {
      return { ...state, ...action.payload };
    }
    default:
      return state;
  }
}

export default followsReducer;
